### PR TITLE
Support Numpy 2.5 and add to CI

### PR DESCRIPTION
**Pull request by seberg**
This includes fixes from gh-10094 plus additional skips, especially the skip-if-after-baseline (which would start failing if we bump the baseline to 2.5+ -- and don't forget to do so in the test).

I have omitted actually adding the CI job here, because this is lower priority then gh-10152 and that'll just be a bunch of merge conflicts.

Closes gh-10094, addresses most of gh-10036

### CODE REVIEW

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 1]
```diff
@@ -548,14 +548,38 @@ def test_shape_mismatch(self):
     {'shape': (3, 4, 5), 'indices': (2, 3), 'out_shape': (2, 3)},
     {'shape': (), 'indices': (), 'out_shape': ()}
 )
-class TestNdarrayTakeErrorTypeMismatch(unittest.TestCase):
+class TestNdarrayTakeTypeMismatch(unittest.TestCase):
+    # NOTE(seberg): Historically cupy was always fully restrictive
+    # while NumPy was just wrong: https://github.com/numpy/numpy/pull/30615
+    # As of CuPy 14.2, CuPy uses 2.5+ (future) behavior.
```
Comment (leofang) [inline tests/cupy_tests/core_tests/test_ndarray.py:558]: Nit: this PR is milestoned for v14.3.0, so this should say "14.3" (or just drop the specific version — the note works fine without pinning to a release).

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 2]
```diff
@@ -548,14 +548,38 @@ def test_shape_mismatch(self):
...
 
-    def test_output_type_mismatch(self):
+    @testing.with_requires('numpy>=2.5')
+    @testing.numpy_cupy_array_equal()
```
Comment (leofang) [inline tests/cupy_tests/core_tests/test_ndarray.py:560]: Because this test is gated on `numpy>=2.5`, the new CuPy relaxation isn't exercised at all on the 2.3 / 2.4 CI cells — i.e. no assertion that CuPy accepts `int64 → int32` (or any other same-kind cast) through `take(out=)` unless the CI matrix happens to include numpy 2.5. Would it be worth a cupy-only smoke assertion so the new behavior is pinned independent of the installed numpy?

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 3]
```diff
@@ -548,14 +548,38 @@ def test_shape_mismatch(self):
...
+    # incorrectly given by numpy (presumably until Deprecation is finalized)
+    @pytest.mark.filterwarnings('ignore::numpy.exceptions.ComplexWarning')
```
Comment (leofang) [inline tests/cupy_tests/core_tests/test_ndarray.py:562]: My bot was saying that casting `int64` to `complex64` genuinely loses precision, so raising `ComplexWarning` here is arguably *correct* rather than "incorrectly."

File: tests/cupy_tests/core_tests/test_ndarray.py (hunk 2 above)
Comment (leofang) [reply tests/cupy_tests/core_tests/test_ndarray.py:560]: Or maybe we should add numpy 2.5 CI in this PR too?

**Review by leofang — COMMENTED** [review]
Thanks, @seberg, mostly LGTM. My main question is if this PR raises the NumPy baseline to v2.5.0? If so, historically it would require a new major release.

File: tests/cupy_tests/core_tests/test_ndarray.py (hunk 3 above)
Comment (seberg) [reply tests/cupy_tests/core_tests/test_ndarray.py:562]: The bot is wrong: `ComplexWarning` is given for complex -> real casts that drop the imaginary part, `real -> complex` don't so never need to warn.

File: tests/cupy_tests/core_tests/test_ndarray.py (hunk 2 above)
Comment (seberg) [reply tests/cupy_tests/core_tests/test_ndarray.py:560]: Yeah, was going to follow up on it, but maybe we might as well include it here.

(And no, this doesn't raise the baseline of course, this is one corner case place where I decided it is a bug fix we might as well do without the raise in CuPy.)

N.B.: The test below with `skip_if_after_baseline` will actually fail when we do (so long we remember to also bump it in the test constant).

**Review by leofang — DISMISSED** [review]
It seems the CI failure is real, otherwise LGTM!

**Comment by seberg** [comment]
Hmmm, yeah, not sure why I had missed that locally.

Note on *minor* change:  CuPy never did NumPy's terrible logic of returning a `eig_result.real` when the imaginary part was all zero *except* for empty/size zero inputs.

That is inconsistent, so I decided that it is best to always go with the NumPy 2.5 behavior even though that isn't yet the base-line.  Special casing size zero arrays isn't useful.

/test mini

File: cupy/linalg/_eigenvalue.py [hunk 4]
```diff
@@ -386,6 +387,7 @@ def eigvals(a):
 
     if a.size == 0:
         _, v_dtype = _util.linalg_common_type(a)
+        v_dtype = numpy.dtype(v_dtype.char.upper())
```
Comment (seberg) [inline cupy/linalg/_eigenvalue.py:390]: I don't like the `char.upper()` way of doing this, but for this PR, also no point in breaking with the files pattern.

**Comment by leofang** [comment]
CI failure seems to be known issue (#10239).

/test force-skip

**Review by leofang — APPROVED** [review]
